core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -414,6 +433,142 @@ private[spark] object ClosureCleaner extends Logging {
   }
 }
 
+private[spark] object IndylambdaScalaClosures extends Logging {
+  // internal name of java.lang.invoke.LambdaMetafactory
+  val LambdaMetafactoryClassName = "java/lang/invoke/LambdaMetafactory"
+  // the method that Scala indylambda use for bootstrap method
+  val LambdaMetafactoryMethodName = "altMetafactory"
+  val LambdaMetafactoryMethodDesc = "(Ljava/lang/invoke/MethodHandles$Lookup;" +
+    "Ljava/lang/String;Ljava/lang/invoke/MethodType;[Ljava/lang/Object;)" +
+    "Ljava/lang/invoke/CallSite;"
+
+  /**
+   * Check if the given reference is a indylambda style Scala closure.
+   * If so, return a non-empty serialization proxy (SerializedLambda) of the closure;
+   * otherwise return None.
+   *
+   * @param maybeClosure the closure to check.
+   */
+  def getSerializationProxy(maybeClosure: AnyRef): Option[SerializedLambda] = {
+    val maybeClosureClass = maybeClosure.getClass
+
+    // shortcut the fast check:
+    // indylambda closure classes are generated by Java's LambdaMetafactory, and they're always
+    // synthetic.
+    if (!maybeClosureClass.isSynthetic) return None
+
+    val implementedInterfaces = ClassUtils.getAllInterfaces(maybeClosureClass).asScala
+    val isClosureCandidate = implementedInterfaces.exists(_.getName == "scala.Serializable") &&
```

Comment:
This detail has [changed](https://github.com/scala/scala/commit/82fd5910a2862ccc9967e99c7e9828453eed195a) in Scala 2.13, in which `scala.Serializable` has become a type alias for `java.io.Serializable`. You could safely add that check in here already to future proof things.

Reply:
Thank you very much! This is very useful information. Does that mean I should check for `maybeClosure .isInstanceOf[java.io.Serializable]` here instead of bothering with `scala.Serializable`? Assuming we're running on a JVM instead of running in a no-Java environment, this instanceof check should always be true for Scala closures, right?

Reply:
Yep, that should work fine.

Scala lambdas can also implement other functional interfaces (such as `java.util.function.Supplier`) when used in a context that expects one. I guess that's out of scope for `ClosureCleaner`, though.